src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -560,6 +564,156 @@ private async ValueTask SendRstStreamAsync(int streamId, Http2ProtocolErrorCode
                 (buffer.Slice(0, FrameHeader.MaxLength), buffer.Slice(FrameHeader.MaxLength)) :
                 (buffer, Memory<byte>.Empty);
 
+        private void WriteHeader(string name, string value)
+        {
+            // TODO: ISSUE 31307: Use static table for known headers
+
+            int bytesWritten;
+            while (!HPackEncoder.EncodeHeader(name, value, _headerBuffer.AvailableSpan, out bytesWritten))
+            {
+                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpan.Length + 1);
+            }
+
+            _headerBuffer.Commit(bytesWritten);
+        }
+
+        private void WriteHeaderCollection(HttpHeaders headers)
+        {
+            foreach (KeyValuePair<HeaderDescriptor, string[]> header in headers.GetHeaderDescriptorsAndValues())
+            {
+                // The Host header is not sent for HTTP2 because we send the ":authority" pseudo-header instead
+                // (see pseudo-header handling below in WriteHeaders)
+                if (header.Key.KnownHeader == KnownHeaders.Host)
+                {
+                    continue;
+                }
+
+                // The Connection header is not supported in HTTP2. Don't send it.
+                if (header.Key.KnownHeader == KnownHeaders.Connection)
+                {
+                    continue;
+                }
+
+                Debug.Assert(header.Value.Length > 0, "No values for header??");
+                for (int i = 0; i < header.Value.Length; i++)
+                {
+                    WriteHeader(header.Key.Name, header.Value[i]);
+                }
+            }
+        }
+
+        private void WriteHeaders(HttpRequestMessage request)
+        {
+            Debug.Assert(_headerBuffer.ActiveMemory.Length == 0);
+
+            // HTTP2 does not support Transfer-Encoding: chunked, so disable this on the request.
+            request.Headers.TransferEncodingChunked = false;
+
+            HttpMethod normalizedMethod = HttpMethod.Normalize(request.Method);
+
+            // TODO: ISSUE 31307: Use static table for pseudo-headers
+            WriteHeader(":method", normalizedMethod.Method);
+            WriteHeader(":scheme", "https");
+
+            string authority;
+            if (request.HasHeaders && request.Headers.Host != null)
+            {
+                authority = request.Headers.Host;
+            }
+            else
+            {
+                authority = request.RequestUri.IdnHost;
+                if (!request.RequestUri.IsDefaultPort)
+                {
+                    // TODO: Avoid allocation here by caching this on the connection or pool
+                    authority += ":" + request.RequestUri.Port;
+                }
+            }
+
+            WriteHeader(":authority", authority);
+            WriteHeader(":path", request.RequestUri.GetComponents(UriComponents.PathAndQuery | UriComponents.Fragment, UriFormat.UriEscaped));
```

Comment:
We want to send the fragment here?

Reply:
I just moved this code (didn't change it). But you're right, this is incorrect. Will fix.

Reply:
The original is correct. This should only send the path and query, and shouldn't include the fragment. ([source](https://http2.github.io/http2-spec/#rfc.section.8.1.2.3))